_posts/2015-04-17-Enums-match-mutation-and-moves.md

```diff
@@ -0,0 +1,636 @@
+contains, and can mutate it, lend out references to it, et cetera.
+When you finish with the box and let it fall out of scope, it will
+automatically clean up the resources associated with the
+heap-allocated `V`.)
+
+The above `enum` definition ensures that if we are given a `BinaryTree`, it
+will always fall into one of the above two cases. One will never
+encounter a `BinaryTree::Node` that does not have a left-hand child.
+There is no need to check for null.
+
+One *does* need to check whether a given `BinaryTree` is a `Leaf` or
+is a `Node`, but the compiler statically ensure such checks are done:
+you cannot accidentally interpret the data of a `Leaf` as if it were a
+`Node`, nor vice versa.
+
+Here is a function that increments all of the integers in a tree
+using `match`.
+
+```rust
+fn tree_weight_v1(t: BinaryTree) -> i32 {
+    match t {
+        BinaryTree::Leaf(payload) => payload,
+        BinaryTree::Node(left, payload, right) => {
+            tree_weight_v1(*left) + payload + tree_weight_v1(*right)
+        }
+    }
+}
+
+/// Returns tree that Looks like:
+///
+///      +----(4)---+
+///      |          |
+///   +-(2)-+      [5]
+///   |     |   
+///  [1]   [3]
+///
+fn sample_tree() -> BinaryTree {
+    let l1 = Box::new(BinaryTree::Leaf(1));
+    let l3 = Box::new(BinaryTree::Leaf(3));
+    let n2 = Box::new(BinaryTree::Node(l1, 2, l3));
+    let l5 = Box::new(BinaryTree::Leaf(5));
+
+    BinaryTree::Node(n2, 4, l5)
+}
+
+#[test]
+fn tree_demo_1() {
+    let tree = sample_tree();
+    assert_eq!(tree_weight_v1(tree), (1 + 2 + 3) + 4 + 5);
+}
+```
+
+Algebraic data types establish structural invariants that are strictly
+enforced by the language. (Even richer representation invariants can
+be maintained via the use of modules and privacy; but let us not
+digress from the topic at hand.)
+
+### Both expression- and statement-oriented
+
+Unlike many languages that offer pattern matching, Rust *embraces*
+both statement- and expression-oriented programming.
+
+Many functional languages that offer pattern matching encourage one to
+write in an "expression-oriented style", where the focus is always on
+the values returned by evaluating combinations of expressions, and
+side-effects are discouraged. This style contrasts with imperative
+languages, which encourage a statement-oriented style that focuses on
+sequences of commands executed solely for their side-effects.
+
+Rust excels in supporting both styles.
+
+Consider writing a function which maps a non-negative integer to a
+string rendering it as an ordinal ("1st", "2nd", "3rd", ...).
+
+The following code uses range patterns to simplify things, but also,
+it is written in a style similar to a `switch` in a statement-oriented
+language like C (or C++, Java, et cetera), where the arms of the
+`match` are executed for their side-effect alone:
+
+```rust
+fn num_to_ordinal(x: u32) -> String {
+    let suffix;
+    match (x % 10, x % 100) {
+        (1, 1) | (1, 21...91) => {
```

Comment:
this code blew my mind somehow. I have never thought about what a frickin' annoying problem it is to generate these suffixes.